File: csharp/LeetCode.Questions/src/Exercises/Solutions/BestTimeToBuyAndSellStock.cs
namespace Exercises.Solutions;
    public class BestTimeToBuyAndSellStock
    {
        public int FindMaxProfit(int[] prices)
        {
            // Inicializamos a variável de lucro máximo como 0 e a variável de preço mínimo com o primeiro elemento do array.
            int maxProfit = 0;
            int minPrice = prices[0];

            // Percorremos os preços das ações a partir do segundo elemento.
            for (int i = 1; i < prices.Length; i++)
            {
                // Capturamos o preço atual.
                int currentPrice = prices[i];

                // Atualizamos o preço mínimo, se o preço atual for menor.
                minPrice = Math.Min(minPrice, currentPrice);

                // Calculamos o lucro máximo ao subtrair o preço atual pelo preço mínimo anterior.
                maxProfit = Math.Max(maxProfit, currentPrice - minPrice);
            }

            // Retornamos o lucro máximo obtido.
            return maxProfit;
        }
    }

File: go/exercises/code/BestTimeToBuyAndSellStock.go
package code

// FindMaxProfit calcula o lucro máximo possível comprando e vendendo uma ação uma única vez.
// Recebe uma lista de preços das ações em dias consecutivos e retorna o lucro máximo.
func FindMaxProfit(prices []int) int {
	// Inicializamos a variável maxProfit como 0 para representar o lucro máximo.
	maxProfit := 0
	// Inicializamos a variável minPrice com o primeiro preço, pois será o menor preço encontrado até o momento.
	minPrice := prices[0]

	// Percorremos os preços das ações a partir do segundo elemento.
	for i := 1; i < len(prices); i++ {
		currentPrice := prices[i] // Capturamos o preço atual.

		// Se o preço atual for menor que o menor preço até o momento, atualizamos minPrice.
		if currentPrice < minPrice {
			minPrice = currentPrice
		}

		// Calculamos o lucro potencial atual (preço atual - menor preço até o momento).
		// Se for maior que o maxProfit registrado, atualizamos maxProfit.
		if currentPrice-minPrice > maxProfit {
			maxProfit = currentPrice - minPrice
		}
	}

	// Retornamos o lucro máximo calculado.
	return maxProfit
}
